File: yt2inv.py
import webbrowser
import os
import shutil

def print_header():
    columns, _ = shutil.get_terminal_size()
    with open("yt2inv.txt", 'r') as file:
        for line in file:
            line = line.rstrip('\n') # remove unneccessary crud
            print(f"{line.center(columns)}")

# print one line strings in the center with delay
def display_center_text(text):
    columns = shutil.get_terminal_size().columns
    centered_text = text.rstrip().center(columns)
    print(centered_text)

def convert_youtube_to_invidious(youtube_url):
    invidious_base_url = "https://yewtu.be"

    if "youtube.com" in youtube_url or "youtu.be" in youtube_url:
        if "watch?v=" in youtube_url:
            video_id = youtube_url.split("watch?v=")[-1]
        elif "youtu.be" in youtube_url:
            video_id = youtube_url.split("youtu.be/")[-1]
        else:
            return "Invalid YouTube URL."

        invidious_url = f"{invidious_base_url}/watch?v={video_id}"
        return invidious_url
    else:
        return "Invalid YouTube URL."

def main():
    os.system('clear')
    display_center_text("yt2inv")
    display_center_text("a youtube to invidious converter")

    history = []

    while True:
        youtube_url = input("\nEnter a YouTube link (or 'q' to quit)\nurl:").strip()

        if youtube_url.lower() in ['q', 'exit']:
            print("Exiting the program. Goodbye!")
            break
        elif youtube_url.lower() == 'history':
            print("\nConverted Links History:")
            for link in history:
                print(link)
            continue

        invidious_link = convert_youtube_to_invidious(youtube_url)

        if invidious_link != "Invalid YouTube URL.":
            history.append(invidious_link)
            print(f"Invidious Link: {invidious_link}")

            open_now = input("Do you want to open this link in your browser now? (y/n): ").strip().lower()
            if open_now == 'y':
                webbrowser.open(invidious_link)
        else:
            print("Invalid YouTube URL. Please make sure the URL is correct and try again.")

if __name__ == "__main__":
    main()

File: yt2inv.cpp
#include <iostream>
#include <string>

std::string convert_youtube_to_invidious(const std::string& youtube_url) {
    const std::string invidious_base_url = "https://yewtu.be/watch?v=";
    std::size_t pos;
    std::string video_id;

    if ((pos = youtube_url.find("youtube.com")) != std::string::npos) {
        pos = youtube_url.find("watch?v=");
        if (pos != std::string::npos) {
            video_id = youtube_url.substr(pos + 8); // 8 is the length of "watch?v="
        }
    } else if ((pos = youtube_url.find("youtu.be")) != std::string::npos) {
        pos = youtube_url.find("youtu.be/");
        if (pos != std::string::npos) {
            video_id = youtube_url.substr(pos + 9); // 9 is the length of "youtu.be/"
        }
    }

    return (!video_id.empty()) ? invidious_base_url + video_id : "Invalid YouTube URL.";
}

int main() {
    std::string youtube_url, invidious_url;

    std::cout << "Welcome to YouTube to Invidious Link Converter!\n";
    std::cout << "Enter a YouTube URL to convert it to an Invidious link:\n";
    std::cout << "Type 'q' to quit.\n";

    while (true) {
        std::cout << "\nEnter a YouTube link: ";
        std::getline(std::cin, youtube_url);

        if (youtube_url == "q") {
            break;
        }

        invidious_url = convert_youtube_to_invidious(youtube_url);

        if (invidious_url != "Invalid YouTube URL.") {
            std::cout << "Invidious Link: " << invidious_url << std::endl;
            // Code to open the URL in a web browser could be added here
        } else {
            std::cout << invidious_url << std::endl;
        }
    }

    std::cout << "Exiting the program. Goodbye!\n";
    return 0;
}
